hoodie-client/src/main/java/com/uber/hoodie/HoodieWriteClient.java

```diff
@@ -776,8 +780,55 @@ public boolean rollback(final String commitTime) throws HoodieRollbackException
     return true;
   }
 
-  private void rollback(List<String> commits) {
-    if (commits.isEmpty()) {
+  /**
+   * NOTE : This action requires all writers (ingest and compact) to a dataset to be stopped before proceeding.
+   * Revert the (inflight/committed) record changes for all commits after the provided @param.
+   * Three steps: (1) Atomically unpublish this commit (2) clean indexing data, (3) clean new generated parquet/log
+   * files and/or append rollback to existing log files. (4) Finally delete .commit, .inflight, .compaction.inflight
+   * or .compaction.requested file
+   */
+  public void restoreToCommit(final String commitTime) throws HoodieRollbackException {
+
+    // Create a Hoodie table which encapsulated the commits and files visible
+    HoodieTable<T> table = HoodieTable.getHoodieTable(
+        new HoodieTableMetaClient(jsc.hadoopConfiguration(), config.getBasePath(), true), config, jsc);
+    // Get all the commits on the timeline after the provided commit time
+    List<HoodieInstant> commitsToRollback = table.getActiveTimeline().getCommitsAndCompactionTimeline().getInstants()
+        .filter(instant -> HoodieActiveTimeline.GREATER.test(instant.getTimestamp(), commitTime))
+        .collect(Collectors.toList());
+    // reverse the commits to descending order of commit time
+    Collections.reverse(commitsToRollback);
+    commitsToRollback.stream().forEach(instant -> {
+      switch (instant.getAction()) {
+        case HoodieTimeline.COMMIT_ACTION:
+        case HoodieTimeline.DELTA_COMMIT_ACTION:
+          rollback(instant.getTimestamp());
+          break;
+        case HoodieTimeline.COMPACTION_ACTION:
+          if (instant.isRequested()) {
+            // TODO : Get file status and create a rollback stat and file
+            // TODO : Delete the .aux files along with the instant file, okay for now since the archival process will
+            // delete these files when it does not see a corresponding instant file under .hoodie
+            deleteRequestedCompaction(instant.getTimestamp());
+            logger.info("Deleted pending scheduled compaction " + instant.getTimestamp());
+          } else {
+            rollback(instant.getTimestamp());
+          }
+          break;
+        default:
+          throw new IllegalArgumentException("invalid action name " + instant.getAction());
+      }
+      try {
```

Comment:
As discussed offline, we can avoid this sleep and have restoreToCommit create only one rollback file for the entire call instead of one rollback per commit getting rolled-back.

Reply:
+1 .. does this need some change to Rollback metadata? would that be backwards compatible

Reply:
Yes, this does change the rollback metadata and makes it backwards incompatible. So I made a small hack in the way the rollback stats are written, will chat f2f about this as well.